firmware/src/temperature.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ * temperature.c
+ * Copyright (C) 2015 Simon Newton
+ */
+
+#include "temperature.h"
+
+#include "system_definitions.h"
+#include "peripheral/adc/plib_adc.h"
+#include "sys/attribs.h"
+
+#include "coarse_timer.h"
+
+#include "app_settings.h"
+
+static const uint16_t SAMPLING_PERIOD = 10000;  // 1s
+
+// The conversion function is:
+//   temp [deci-degrees] = m * sampled_value + c
+//
+// The MCP9701AT device has:
+//  - 400mV @ 0 degrees C
+//  - 19.5mV per 1 degree
+//
+// With 10-bit sampling and a 3v3 ref we have:
+//   temp = ((3300 / 1024 * sample - 400) / 19.5 * 10
+//
+// Which reduces to:
+//   temp = 1.6526 * sample - 205.128;
+static const float CONVERSION_MULTIPLIER = 1.6526;
+static const float CONVERSION_OFFSET = -205.128;
+
+static CoarseTimer_Value g_timer;
+
+struct {
+  bool new_sample;  //!< true if there is a new sample
+  uint16_t sample_value;  //!< The raw sampled value
+  uint16_t temperature;  //!< The temperature in 10ths of a degree.
+} g_adc_data;
+
+void __ISR(_ADC_VECTOR, ipl1AUTO) ADCEvent() {
+  // AD1CON1bits.ASAM = 0;
```

Comment:
The ADC has a number of different modes and requires multiple registers to be configured. I've written out the config values here so when we next have to change the code we know what's going on. 

Between 1.02 and 1.06 some of the harmony function names changed. Being able to map to registers is useful.